### ISSUE TITLE
"During interactive linking, GHCi couldn't find the following symbol" with -fno-code

### ISSUE THREAD

**Issue by mrkgnao** [ndmitchell/ghcid#101]
I attempted to use `ghcid` with the `yi-all-static` config of Yi.

```
> ghcid -c "stack ghci --ghci-options -fno-code"          11.5s < Sun 30 Jul 2017 04:03:57 PM IST
Loading stack ghci --ghci-options -fno-code ...
The following GHC options are incompatible with GHCi and have not been passed to it: -threaded
Configuring GHCi with the following packages: yi-all-static, yi-core, yi-frontend-pango, yi-frontend-vty, yi-keymap-emacs, yi-keymap-vim, yi-language, yi-misc-modes, yi-mode-haskell, yi-mode-javascript
Using main module: 1. Package `yi-all-static' component exe:yi with main-is file: /home/mrkgnao/code/haskell/yi/example-configs/yi-all-static/Main.hs
GHCi, version 8.0.2: http://www.haskell.org/ghc/  :? for help
[ 1 of 81] Compiling Yi.Utils         ( /home/mrkgnao/code/haskell/yi/yi-language/src/Yi/Utils.hs, nothing )

[...]

[20 of 81] Compiling Yi.Buffer.Implementation ( /home/mrkgnao/code/haskell/yi/yi-core/src/Yi/Buffer/Implementation.hs, nothing )

ByteCodeLink.lookupCE
During interactive linking, GHCi couldn't find the following symbol:
  YiziUtils_makeLensesWithSuffix_closure
This may be due to you not asking GHCi to load extra object files,
archives or DLLs needed by your current session.  Restart GHCi, specifying
the missing library using the -L/path/to/object/dir and -lmissinglibname
flags, or simply by naming the relevant files on the GHCi command line.
Alternatively, this link failure might indicate a bug in GHCi.
If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
```

Is this expected? Without `-fno-code`, or using `-fobject-code` instead, the problem goes away.

**Comment by Saizan** [comment]
it's because ```Yi.Utils``` defines the ```makeLensesWithSuffix``` macro, which needs to actually run when called from ```Yi.Buffer.Implementation``` so the code is indeed needed.

it would be great if ghc supported -fobject-code on a per-module basis, so we could use it only on these macro-defining modules but it doesn't seem to.

**Comment by Saizan** [comment]
Actually it seems it's going to be fixed in 8.4.1

https://ghc.haskell.org/trac/ghc/ticket/8025

**Comment by ndmitchell** [comment]
Awesome that it will be fixed soon. Until then I should probably add a note that you should not use `-fno-code` with TemplateHaskell if you define your own TemplateHaskell.

**Comment by ChShersh** [comment]
I'm having the same issue with custom TemplateHaskell macro for lenses. Currently I'm running `ghcid` with next command (and everything works).

```bash
ghcid -c "stack repl"
```

**Comment by ndmitchell** [comment]
The README now says:

> * _I get "During interactive linking, GHCi couldn't find the following symbol"._ This problem is a manifestation of [GHC bug 8025](https://ghc.haskell.org/trac/ghc/ticket/8025), which is fixed in GHC 8.4 and above. Ghcid automatically appends `-fno-code` to the command line, but for older GHC's you can supress that with `--test "return ()"` (to add a fake test) or `-c "ghci ..."` to manually specify the command to run.

Note that if you pass `-c` then `ghcid` doesn't guess which flags it should apply, and thus doesn't guess at `-fno-code`.
